lib/latte.ts

```diff
@@ -0,0 +1,482 @@
+  }
+  cleanup: () => void
+  // The head and worker each define seperate onTest handlers
+  onTest?: (test: Test, err: unknown, log: string[]) => void
+}
+
+type describe = {
+  (name: string, fn: () => void): void
+  skip: () => void
+  only: () => void
+}
+
+declare global {
+  interface Window {
+    Latte: Latte
+    describe: describe
+    context: describe
+    timeout: () => void
+    beforeEach: TestSetupFn
+    afterEach: TestSetupFn
+    before: TestSetupFn
+    after: TestSetupFn
+    helper: (name: string, cb: FnOrGroup) => void
+    it: {
+      (name: string, cb?: TestFn): void
+      skip: () => void
+      only: () => void
+    }
+    setup: (opts: LatteOptions) => void
+  }
+}
+
+// TODO refactor latte into a class, this will make the typing way easier
+
+// Right now there is nothing to export, but TS requires this to do the editing of Window
+(function () {
+  let mode: 'headless' | 'debug' // headless or debug
+  let willHotReload = false
+  let currentContext: null | TestContext = null // The context of the currently running test.
+  let currentTimeout: number
+
+  let whenCurrentTestFinished: Promise<void> | null = null // promise that's resolved when the currently running test finishes
+  let currentTestResolve: null | (() => void) = null
+
+  // Save some globals in case test code overrides it (ie sinon, bluebird, etc)
+  const realSetTimeout = window.setTimeout
+  const realClearTimeout = window.clearTimeout
+  const RealPromise = window.Promise
+
+  // polyfill promise.finally without changing the prototype
+  const promiseFinally =
+    RealPromise.prototype.finally ||
+    function <T>(this: Promise<T>, fn: () => Promise<void>) {
+      return this.then(
+        (val: T) => Promise.resolve(fn()).then(() => val),
+        (err: Error) =>
+          Promise.resolve(fn()).then(() => {
+            throw err
+          })
+      )
+    }
+
+  const Latte: Latte = {
+    root: {
+      children: [],
+      tests: [],
+      depth: 0,
+      testCount: 0,
+      before: [],
+      beforeEach: [],
+      after: [],
+      afterEach: [],
+      helpers: {},
+    },
+    setup: function latteSetup(opts) {
+      mode = opts.mode
+      willHotReload = opts.willHotReload
+    },
+    waitForCurrentTest: async function waitForCurrentTest() {
+      whenCurrentTestFinished && (await whenCurrentTestFinished)
+    },
+    flatten: function latteFlatten(suite: TestSuite = window.Latte.root) {
+      let allTests: Test[] = []
+      let queue = [suite]
+
+      let step = queue.pop()
+      while (step) {
+        allTests = allTests.concat(step.tests)
+        queue = queue.concat(step.children)
+
+        step = queue.pop()
+      }
+      return allTests
+    },
+    run: async function latteRun(tests) {
+      // Store logs so they can be relayed back from remote
+      const log: string[] = []
+      const originalConsoleLog = console.log
+      console.log = (...args) => {
+        log.push(args.join(' '))
```

Comment:
No changes needed... 
Can we make this work for non-string logs?

Reply:
The next step outside of this file is to JSON.stringify this content. It is probably not worth it right now to make sure that all looks good, I'll make it better if there is an issue where that will help with